### PR TITLE
App to router

### PR DESCRIPTION
**Pull request by jacargentina**
This PR makes possible this use case: having just 1 keystone instance on the same node process, and use `express-vhost` to serve multiple subdomains on it.

The key to this is the new `keystone.initRouter` which basically _mounts_ keystone middleware and admin ui on a given express router, while also passing a router binding function for custom routes.

This code is an example on how i'm using it on my own setup:

``` javascript
var evh = require('express-vhost'),
    keystone = require('keystone'),
    express = require('express');

var server = express();
keystone.initExpressApp(server);
keystone.openDatabaseConnection();
server.use(evh.vhost(server.enabled('trust proxy')));

var www = express.Router();
keystone.initRouter(www, require('./routes/www'));

var erp_product_minisite = express.Router();
keystone.initRouter(erp_product_minisite, require('./routes/erp'));

evh.register('www.example.com', www);
evh.register('erp.example.com', erp_product_minisite);

// keystone is a proxy for a Gitlab instance also
var gitOptions = url.parse('http://127.0.0.1:8081/');
gitOptions.preserveHost = true;

evh.register('gitlab.example.com', proxy(gitOptions));

server.listen();
```

For using keystone as usual, the  `keystone.start` method now uses a new `keystone.initDefaultRouter` for mounting a default router on the / of `keystone.app`

This is really useful for any setup where you can have the same keystone admin UI, same models, etc; but having many _subsites_

### CODE REVIEW

**Comment by r3wt** [comment]
Very cool! :+1:

**Comment by webteckie** [comment]
LRGTM cc: @JedWatson

**Comment by jacargentina** [comment]
@jedwatson this is the complement to the cookie signin PR. please evaluate it.

**Comment by jacargentina** [comment]
@jedwatson happy 3 months for this PR. It is enough waiting ? 😂

**Comment by wmertens** [comment]
@jacargentina can you rebase?

**Comment by wmertens** [comment]
I'm not convinced that renaming `app` to `router` makes much difference. This PR would be a lot smaller and less breaking other PRs if you don't rename app to router everywhere. You could have `app` mean "the current app" and  `keystone.app` would be `keystone.adminApp`?

Also, your renaming/moving the `createApp.js` file makes it hard to see what changed, you should do that in 2 commits if you really want to do that, one for the rename and one for the changes.

I like that this PR reduces internal coupling but it should be more concise.

**Comment by jacargentina** [comment]
@wmertens I get the pat for dont renaming app to router, but the `createApp.js` deleting is an atomic replace for the new ìnitDefaultRouter.js` so i prefer to leave it as is.

**Comment by wmertens** [comment]
Well as it is now, there is code shaoed between those two files but git
can't tell us what it is and that makes reviewing harder…

**Comment by jacargentina** [comment]
Ok. Ill try to make a rename lije u want

File: server/createApp.js [hunk 1]
```diff
@@ -4,136 +4,115 @@ var methodOverride = require('method-override');
...
-			for (var key in loggerOptions.tokens) {
-				if (hasOwnProperty.call(loggerOptions.tokens, key) && typeof loggerOptions.tokens[key] === 'function') {
-					morgan.token(key, loggerOptions.tokens[key]);
-				}
-			}
-		}
-
-		app.use(morgan(keystone.get('logger'), loggerOptions));
+	if (this.get('logger')) {
+		router.use(morgan(this.get('logger'), this.get('logger options')));
```
Comment (wmertens) [inline server/createApp.js]: You moved the morgan use before the 'pre:logger' hook. Intentional?

File: server/createApp.js [hunk 2]
```diff
@@ -4,136 +4,115 @@ var methodOverride = require('method-override');
...
-			express = require('express');
-		}
-		keystone.app = express();
-	}
-
-	var app = keystone.app;
-
-	keystone.initDatabase();
-	keystone.initExpressSession(keystone.mongoose);
+	var keystone = this;
```
Comment (wmertens) [inline server/createApp.js]: I don't understand why you assign `this` to `keystone` and then rename `keystone` almost everywhere to `this`?

File: server/createApp.js (hunk 1 above)
Comment (wmertens) [inline server/createApp.js]: Also, why remove all that morgan init code?

File: server/createApp.js [hunk 3]
```diff
@@ -4,136 +4,115 @@ var methodOverride = require('method-override');
...
+
 	// unless the headless option is set (which disables the Admin UI),
 	// bind the Admin UI's Dynamic Router
-	if (!keystone.get('headless')) {
-		app.use(function (req, res, next) {
-			keystone.callHook('pre:admin', req, res, next);
-		});
-		app.use('/' + keystone.get('admin path'), require('../admin/server').createDynamicRouter(keystone));
+	if (!this.get('headless')) {
+		router.use('/' + this.get('admin path'), require('../admin/server').createDynamicRouter(this));
```
Comment (wmertens) [inline server/createApp.js]: why remove the 'pre:admin' hook?

File: server/createApp.js [hunk 4]
```diff
@@ -4,136 +4,115 @@ var methodOverride = require('method-override');
...
-	app.use(function (req, res, next) {
+	router.use(function (req, res, next) {
 		keystone.callHook('pre:routes', req, res, next);
 	});
 
-	// Configure application routes
-	if (typeof keystone.get('routes') === 'function') {
-		keystone.get('routes')(app);
+	// Configure React routes
+	if (this.get('react routes')) {
```
Comment (wmertens) [inline server/createApp.js]: I have the impression you copied an old version of createApp… react routes is gone iirc…

File: server/initViewLocals.js [hunk 5]
```diff
@@ -1,14 +1,13 @@
...
+module.exports = function initViewLocals (keystone) {
 	// Apply locals
 	if (typeof keystone.get('locals') === 'object') {
-		assign(app.locals, keystone.get('locals'));
+		assign(keystone.app.locals, keystone.get('locals'));
 	}
 
 	// Default "pretty html" mode except in production
-	// Only if it has not been specified in the locals setting
-	if (app.locals.pretty === undefined && keystone.get('env') !== 'production') {
```
Comment (wmertens) [inline server/initViewLocals.js]: are you intentionally removing this condition?

**Comment by jacargentina** [comment]
@wmertens i dont understand the travis error.... assign not used. I've not changed anything related to fields?!

**Comment by JedWatson** [comment]
@jacargentina can you review the three ways of setting up keystone w/ express in the [test project](https://github.com/keystonejs/keystone-test-project) (see `keystone.js`, `express.js` and `advanced.js`) and submit a PR illustrating the _public_ API you want to use here?

I'm stuck working out what this specifically does that you can't do already, and wondering if there's a simpler change we can make to enable it. It's also hard to work out how those existing configuration methods would break/change with what you're proposing here. Seeing them side by side would be helpful.

**Comment by wmertens** [comment]
@jacargentina the travis error is a lint error outside of your PR. If you rebase on latest master it is probably fixed.

For the rest the PR looks good! 👍 on the technical side of the PR, but I am also curious about the use cases :)

**Comment by jacargentina** [comment]
@JedWatson the PR for https://github.com/keystonejs/keystone-test-project is ready
Hope it helps understand my very complex use case (many subdomains, all served on the same express instance). Also i'm having a gitlab proxied on it.

**Comment by jacargentina** [comment]
It is rebased on todays upstream master: i dont know why the Node 5 tests failed

**Comment by jacargentina** [comment]
Getting old again? Please, took lots of time and effort to rebase... I'm sadly losing interest on helping with coding because of PR like this

**Comment by mxstbr** [comment]
Sorry @jacargentina, the reason why this is being "ignored" is that we want to get a beta of Keystone 4 out on npm this or next week!

We're hard at work making that happen, and since this isn't a critical change for this release this'll have to wait until we're done in an effort of finally getting v4 out of the door! I hope you understand 😊

**Comment by JedWatson** [comment]
Sorry for leaving this so long without review or merge. Since it's out of date and in conflict, I'm going to close it.

**Comment by jacargentina** [comment]
@jedwatson nevermind ... I've given up on this project long ago. Went to strapi, CRA. Way better support! Cheers

**Pull request closed without merging**